File: src/main/java/com/dotronglong/faker/service/SourceWatcher.java
package com.dotronglong.faker.service;

import com.dotronglong.faker.contract.Watcher;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.nio.file.*;
import java.util.List;
import java.util.function.BiConsumer;

import static java.nio.file.StandardWatchEventKinds.OVERFLOW;

@Service
public class SourceWatcher implements Watcher {
    private static final Logger logger = LoggerFactory.getLogger(SourceWatcher.class);

    @Override
    public void watch(Path folder, List<WatchEvent.Kind<?>> events, BiConsumer<Path, WatchEvent.Kind<?>> consumer) {
        Thread thread = new Thread(() -> {
            try {
                WatchService watcher = FileSystems.getDefault().newWatchService();
                folder.register(watcher, (WatchEvent.Kind<?>[]) events.toArray());
                logger.info("Start watching {} ...", folder.toString());

                for (;;) {
                    WatchKey key;
                    try {
                        key = watcher.take();
                    } catch (InterruptedException e) {
                        logger.warn("Unable to retrieve watch key. Exception: {}", e.getMessage());
                        return;
                    }

                    for (WatchEvent<?> event: key.pollEvents()) {
                        WatchEvent.Kind<?> kind = event.kind();

                        if (kind == OVERFLOW) {
                            continue;
                        }

                        @SuppressWarnings("unchecked") WatchEvent<Path> ev = (WatchEvent<Path>) event;
                        consumer.accept(folder.resolve(ev.context()), kind);
                    }

                    if (!key.reset()) {
                        break;
                    }
                }
            } catch (IOException e) {
                logger.error("Unable to watch source. Exception: {}", e.getMessage());
            }
        });
        thread.start();
    }
}

File: src/main/kotlin/com/dotronglong/faker/service/watcher/SourceWatcher.kt
package com.dotronglong.faker.service.watcher

import com.dotronglong.faker.contract.Watcher
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.io.IOException
import java.nio.file.FileSystems
import java.nio.file.Path
import java.nio.file.StandardWatchEventKinds.OVERFLOW
import java.nio.file.WatchEvent
import java.nio.file.WatchKey
import java.util.function.BiConsumer

@Service
class SourceWatcher : Watcher {
    private val logger: Logger = LoggerFactory.getLogger(javaClass)

    override fun watch(folder: Path, events: List<WatchEvent.Kind<*>>, consumer: BiConsumer<Path, WatchEvent.Kind<*>>) {
        Thread {
            try {
                val watcher = FileSystems.getDefault().newWatchService()
                folder.register(watcher, events.toTypedArray())
                logger.info("Start watching {} ...", folder.toString())
                while (true) {
                    var key: WatchKey
                    key = try {
                        watcher.take()
                    } catch (e: InterruptedException) {
                        logger.warn("Unable to retrieve watch key. Exception: {}", e.message)
                        break
                    }
                    for (event in key.pollEvents()) {
                        val kind = event.kind()
                        if (kind === OVERFLOW) {
                            continue
                        }

                        @Suppress("UNCHECKED_CAST")
                        consumer.accept(folder.resolve((event as WatchEvent<Path>).context()), kind)
                    }
                    if (!key.reset()) {
                        break
                    }
                }
            } catch (e: IOException) {
                logger.error("Unable to watch source. Exception: {}", e.message)
            }
        }.start()
    }
}